### PR TITLE
Add buffer pooling to JsonResult

### PR DESCRIPTION
**Pull request by rynowak**
/cc @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Formatters.Json/Infrastructure/JsonResultExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+using Microsoft.Extensions.OptionsModel;
+using Microsoft.Net.Http.Headers;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNet.Mvc.Infrastructure
+{
+    /// <summary>
+    /// Executes a <see cref="JsonResult"/> to write to the response.
+    /// </summary>
+    public class JsonResultExecutor
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/Infrastructure/JsonResultExecutor.cs]: This is mostly copy-paste from `JsonResult`

File: test/Microsoft.AspNet.Mvc.Formatters.Json.Test/Infrastructure/JsonResultExecutorTest.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+using Microsoft.Net.Http.Headers;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Infrastructure
+{
+    public class JsonResultExecutorTest
+    {
+        [Fact]
+        public async Task ExecuteAsync_UsesDefaultContentType_IfNoContentTypeSpecified()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Formatters.Json.Test/Infrastructure/JsonResultExecutorTest.cs]: Tests are cut-paste

File: test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonResultTest.cs [hunk 3]
```diff
@@ -21,155 +18,24 @@ namespace Microsoft.AspNet.Mvc
...
-        {
-            // Arrange
-            var expected = _abcdUTF8Bytes;
-
-            var context = GetHttpContext();
-            context.Response.ContentType = responseContentType;
-            var actionContext = new ActionContext(context, new RouteData(), new ActionDescriptor());
-
-            var result = new JsonResult(new { foo = "abcd" });
+            var result = new JsonResult(value);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonResultTest.cs]: Leaving one test behind to verify the end to end. The details are tested in the tests for the executor.

File: src/Microsoft.AspNet.Mvc.Formatters.Json/Infrastructure/JsonResultExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+            if (result.StatusCode != null)
+            {
+                response.StatusCode = result.StatusCode.Value;
+            }
+
+            var serializerSettings = result.SerializerSettings ?? Options.SerializerSettings;
+
+            Logger.JsonResultExecuting(result.Value);
+            using (var writer = WriterFactory.CreateWriter(response.Body, resolvedContentTypeEncoding))
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/Infrastructure/JsonResultExecutor.cs]: Here's the only real semantic change in this PR.

**Comment by pranavkm** [comment]
:shipit:

**Pull request closed without merging**